src/dialogxml/dialogs/strchoice.cpp

```diff
@@ -156,6 +162,88 @@ bool cStringChoice::onOkay(cDialog& me){
 	return true;
 }
 
+bool cStringChoice::onSearch(cDialog& me){
+	cLed& reverse_led = dynamic_cast<cLed&>(me["reverse"]);
+	bool reversed = reverse_led.getState() == led_red;
+	size_t page_delta = reversed ? -1 : 1;
+	size_t loop_from_page = reversed ? 0 : lastPage();
+	size_t loop_to_page = !reversed ? 0 : lastPage();
+	std::string loop_from_str = reversed ? "beginning" : "end";
+	std::string loop_to_str = !reversed ? "beginning" : "end";
+	size_t start_page = page;
+	bool looped_once = false;
+	std::string new_search = me["search-field"].getText();
+	boost::to_lower(new_search);
+	bool repeat_search = new_search == search_str;
+	search_str = new_search;
+	cControl& output = me["search-label"];
+
+	clearHighlights();
+
+	if(search_str.empty()) {
+		output.setText("");
+		return true;
+	}
+
+	bool found_next = false;
+	if(!repeat_search){
+		// First-time search, check the current page before paging forward
+		if(highlightSearch()) found_next = true;
+	}
+
+	while(!found_next){
+		if(page == loop_from_page){
+			// Looped already, found nothing. Avoid infinite loop:
+			if(looped_once){
+				break;
+			}
+			output.setText("Reached the " + loop_from_str + ". Starting from the " + loop_to_str);
```

Comment:
I know we don't have localization yet, and plenty of stuff is in an unlocalizable state, but it would be nice if we can avoid adding _more_ stuff that's unlocalizable. In this case, the thing to avoid is subsentence string concatenation. I don't think it's difficult to simply set `loop_from_str` and `loop_to_str` to just contain the whole sentence? Then this would become `output.setText(loop_from_str + " " + loop_to_str")`.